### PR TITLE
Add some expression functions

### LINKED ISSUE #1304: [FEATURE] Expression functions for offset_curve and single_sided_buffer
Original commit: https://github.com/qgis/QGIS/commit/298d04712594ebe0255266910771a3c735987fb4 by nyalldawson

Especially useful with geometry generators!

### LINKED ISSUE #1348: [FEATURE] support aggregation of geometry
Original commit: https://github.com/qgis/QGIS/commit/2a326ef8adb0769a012cae999a668501042ace38 by nirvn

This feature adds a 'collect' aggregation method resulting in asingle multipart geometry from a list of geometries. This is exposedin the expression engine via the existing aggregate() function,as well as a new collect() function.

### LINKED ISSUE #1379: [FEATURE] New simplify + smoothing expression functions
Original commit: https://github.com/qgis/QGIS/commit/679797e22bcc3d29840248c9ec73fc2911c6cb40 by nyalldawson

Exposes simplification and smoothing algorithms to expression
engine, via:
- simplify(): applies Douglas-Peucker geometry simplification
- simplify_vw(): applies Visvalingam-Whyatt geometry simplification
- smooth(): smoothes a geometry

Carto tip: use smooth along with geometry generators to minimise
the typical "GIS" noded look of rendered geometries!

### LINKED ISSUE #1472: [FEATURE] array support for the replace() expression function
Original commit: https://github.com/qgis/QGIS/commit/8b74201470d98eb02871211f510a1edcaf6980fc by nirvn (https://github.com/qgis/QGIS/pull/3670)

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #1475: [FEATURE] Add is_selected and num_selected functions
Original commit: https://github.com/qgis/QGIS/commit/01f3c9ae38de51e82a4a18a48488695deb51f72f by m-kuhn
- is_selected() returns if the current feature is selected
- num_selected() returns the number of selected features on the current layer
- is_selected(layer, feature) returns if the "feature" is selected. "feature"
  must be on "layer".
- num_selected(layer) returns the number of selected features on "layer"

### LINKED ISSUE #1476: [FEATURE] add string_to_array() and array_to_string() functions
Original commit: https://github.com/qgis/QGIS/commit/7373036edf6e9e549a0871f6f778f13b7f6902d8 by nirvn (https://github.com/qgis/QGIS/pull/3678)

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #1479: [FEATURE] add array_distinct() function
Original commit: https://github.com/qgis/QGIS/commit/17fc7dbe53367f66c0b68930b6d278e41d896b49 by nirvn (https://github.com/qgis/QGIS/pull/3685)

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #1486: [FEATURE] New expression function 'extend'
Original commit: https://github.com/qgis/QGIS/commit/2b545822e7755fe3ca92c1b90db8fe260e40dc30 by nyalldawson

Extends linestrings by a specified amount at the start and
end of the line

### LINKED ISSUE #1509: [FEATURE] add regexp_matches() function
Original commit: https://github.com/qgis/QGIS/commit/1d245b2ef63eb62b40715d63b28b015f20df0d26 by nirvn

The new function returns an array of strings captured by capturinggroups in a supplied regular expression. For e.g., the followingexpression: regexp_matches('qgis=>rocks','(.*)=>(.*)')will return the following array: 'qgis', 'rocks'... :-(

### LINKED ISSUE #1514: [FEATURE][expression] strpos() and regexp_match() improvements
Original commit: https://github.com/qgis/QGIS/commit/6b2b4c578ac45ae6f5745c240e03265c14cdfde1 by nirvn

- strpos() now relies on a simple string within a string search
- regexp_match() now returns pos of a matching regular expression

### LINKED ISSUE #1521: [FEATURE] pole_of_inaccessibility expression function
Original commit: https://github.com/qgis/QGIS/commit/1f81a7c4fc2279f895a91dd3e694730c1bd02704 by nyalldawson

Exposes calculation of pole of inaccessiblity to expression
engine

### LINKED ISSUE #1539: [FEATURE] Expression variables for project CRS
Original commit: https://github.com/qgis/QGIS/commit/188033a6afc298864860ba94a3f150dd5e23f86d by nyalldawson

Adds @project_crs and @project_crs_definition variables
for retrieving the current project CRS

### LINKED ISSUE #1540: [FEATURE] raster_statistic expression function for retrieving
raster band stats from a loaded layer
Original commit: https://github.com/qgis/QGIS/commit/20dc7fb266efba2f2bddc47443f129bf3eef9f03 by nyalldawson

Allows raster band stats (eg min, max, avg) to be used in
expressions

### LINKED ISSUE #1615: [FEATURE] new global expression variable $qgis_locale
Original commit: https://github.com/qgis/QGIS/commit/c2bd4b940bc8ed76f106f72126aa59bd711cc80d by alexbruy

Returns current locale used by QGIS. By default this is current
system locale. If user enabled locale override in QGIS settings
overriden locale will be returned.

### LINKED ISSUE #1616: [FEATURE] item_variables expression function inside compositions
Original commit: https://github.com/qgis/QGIS/commit/f2032ea2688715163a631cc0860b9e5f97257cfc by nyalldawson

This adds a new item_variables expression function when expressions
are used inside a composition context.

The function takes a single argument, the id for an item inside
the composition, and returns a map of variable name to value
for that item.

This allows you to do things like insert text in a label fetching
properties of another item in the composition, eg

Insert scale of map into a label:

map_get( item_variables( 'map'),'map_scale')

Insert x coordinate of map center into a label:

x(map_get( item_variables( 'map'),'map_extent_center'))

### LINKED ISSUE #1617: [FEATURE] New expression variables for map settings
Original commit: https://github.com/qgis/QGIS/commit/db1009cbfa4b83c9b44f09a50928e720645d75cf by nyalldawson

- map_crs, map_crs_definition: retrieves crs details for map
- map_units: retrieves units for map (eg 'meters','degrees')

### LINKED ISSUE #1649: [FEATURE] expression engine's is_layer_visible() (#4045)
Original commit: https://github.com/qgis/QGIS/commit/ecf1f5a5762b4b5e4e188efe9141dfb9574ac0a9 by web-flow

Unfortunately this naughty coder did not write a description... :-(

### LINKED ISSUE #1715: [FEATURE] epoch() expression function (#4151)
Original commit: https://github.com/qgis/QGIS/commit/07cbfdd2e3816cb416f25e6066f689d464676668 by web-flow

This commit adds an epoch() function to the expression engine which
returns the interval in milliseconds between the unix epoch and a
given date value.

### LINKED ISSUE #1723: [FEATURE] Add new expression function `env`
Original commit: https://github.com/qgis/QGIS/commit/26557c9a88e8a74e308a414f8cfa9d445b8df71a by m-kuhn

Unfortunately this naughty coder did not write a description... :-(

### PR DESCRIPTION
**Pull request by DelazJ**
As usual the first commit (@fd50093) is compatible with 2.18
I let some comments because of missing description in the application (and reported the bug on hub)
Fixes #1723, Fixes #1715, Fixes #1649, Fixes #1617, Fixes #1616, Fixes #1615, Fixes #1540, Fixes #1539, Fixes #1521, Fixes #1514, Fixes #1509, Fixes #1486, Fixes #1479, Fixes #1476, Fixes #1475, Fixes #1472, Fixes #1379, Fixes #1348,  Fixes #1304

### CODE REVIEW

**Comment by DelazJ** [comment]
I fail to find what makes Travis complain. Any help?

File: source/docs/user_manual/working_with_vector/expression.rst [hunk 1]
```diff
@@ -337,6 +337,42 @@ This group contains functions which aggregate values over layers and fields.
      expression := "field_from_related_table")
 
 
+Array Functions
+---------------
+
+.. missing description:
+ This group contains functions to ....
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/expression.rst]: This group contains functions to create and manipulate arrays?

File: source/docs/user_manual/working_with_vector/expression.rst [hunk 2]
```diff
@@ -488,12 +547,17 @@ This group  contains general assorted functions.
 ====================  =======================================================
  Function              Description
 ====================  =======================================================
+ env                   Gets an environment variable and returns its content
+                       as a string. If the variable is not found, ``NULL``
+                       will be returned.
  eval                  Evaluates an expression which is passed in a string.
                        Useful to expand dynamic parameters passed as context
                        variables or fields
+ is_layer_visible      Returns true if a specified layer is visible
```
Comment (SrNetoChan) [inline source/docs/user_manual/working_with_vector/expression.rst]: True?

**Comment by SrNetoChan** [comment]
LGTM

File: source/docs/user_manual/working_with_vector/expression.rst (hunk 2 above)
Comment (DelazJ) [inline source/docs/user_manual/working_with_vector/expression.rst]: looks like the lower case is used in expression string builder dialog

**Comment by DelazJ** [comment]
> Great work splitting 2.18 and 3.0 features

`git add -p .` is your friend :-)

**Comment by SrNetoChan** [comment]
Have you seen `git add -i` It's also very cool.

**Comment by DelazJ** [comment]
It looks like `-p` is a subcommand of `-i`. Is this PR mergeable?
I guess it would need to manually close the issues because the target branch (master) is not the default branch of the repo. In which case, is that possible for a maintainer to temporarily switch default branches, merge and restore default? /*me lazy to browse 19 issues to close*/

**Comment by SrNetoChan** [comment]
I think you can merge it. And I would it will fix the issues anyway. Issues are not tied to a specific branch, I think.

**Comment by DelazJ** [comment]
> it will fix the issues anyway. Issues are not tied to a specific branch, I think.

Not sure, because of https://github.com/blog/1506-closing-issues-via-pull-requests and that I had to manually close issues that my recent master-related PRs were supposed to fix.
Anyway, I'll do that this evening when i have a git infrastructure to backport the 2.18 commit at the same time, and if nobody granted my request in the meantime.

**Comment by DelazJ** [comment]
As I suspected, I had to close them one by one manually :(

**Pull request merged**